src/artifacts/artifacts.service.ts

```diff
@@ -0,0 +1,176 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import crypto from 'node:crypto';
+
+import { Loaded, ref } from '@mikro-orm/core';
+import dayjs from 'dayjs';
+
+import { ArtifactCreateBody, ArtifactCreateResponse } from './dtos/artifact-create.js';
+import type { Artifact as ArtifactDto } from './dtos/artifact.js';
+import { Artifact, ArtifactType } from './entities/artifact.entity.js';
+import { ArtifactReadParams, ArtifactReadResponse } from './dtos/artifact-read.js';
+import {
+  ArtifactUpdateBody,
+  ArtifactUpdateParams,
+  ArtifactUpdateResponse
+} from './dtos/artifact-update.js';
+import { ArtifactDeleteParams, ArtifactDeleteResponse } from './dtos/artifact-delete.js';
+import { AppArtifact } from './entities/app-artifact.entity.js';
+import {
+  ArtifactSharedReadParams,
+  ArtifactSharedReadQuery,
+  ArtifactSharedReadResponse
+} from './dtos/artifact-shared-read.js';
+import { ArtifactsListQuery, ArtifactsListResponse } from './dtos/artifacts-list.js';
+import { ArtifactShared } from './dtos/artifact-shared.js';
+
+import { Thread } from '@/threads/thread.entity.js';
+import { Message } from '@/messages/message.entity.js';
+import { APIError, APIErrorCode } from '@/errors/error.entity.js';
+import { createDeleteResponse } from '@/utils/delete.js';
+import { getUpdatedValue } from '@/utils/update.js';
+import { createPaginatedResponse, getListCursor } from '@/utils/pagination.js';
+import { ORM } from '@/database.js';
+
+export function toDto(artifact: Loaded<Artifact>): ArtifactDto {
+  return {
+    ...toSharedDto(artifact),
+    object: 'artifact',
+    thread_id: artifact.thread.id,
+    message_id: artifact.message.id
+  };
+}
+
+export function toSharedDto(artifact: Loaded<Artifact>): ArtifactShared {
+  const dto = {
+    id: artifact.id,
+    object: 'artifact.shared' as const,
+    type: artifact.type,
+    metadata: artifact.metadata ?? {},
+    created_at: dayjs(artifact.createdAt).unix(),
+    source_code: (artifact as AppArtifact).sourceCode,
+    share_url: artifact.accessSecret
+      ? `/v1/artifacts/${artifact.id}/shared?secret=${artifact.accessSecret}`
+      : '',
+    name: artifact.name,
+    description: artifact.description ?? ''
+  };
+  switch (artifact.type) {
+    case ArtifactType.APP:
+      return {
+        ...dto,
+        source_code: (artifact as AppArtifact).sourceCode
+      };
+  }
+}
+
+function getSecret() {
+  return crypto.randomBytes(24).toString('base64url');
+}
+
+export async function createArtifact(body: ArtifactCreateBody): Promise<ArtifactCreateResponse> {
```

Comment:
Do we want to allow users to create artifacts? Shouldn't artifacts come from within the system?

Reply:
This is a good question. but the artifact can be part of the message so we should somehow index this part. We can discuss this.